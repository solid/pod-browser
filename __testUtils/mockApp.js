/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  createThing,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, schema } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/addressBook";

export const oidcPrefix = "http://www.w3.org/ns/solid/oidc#";
export const TOS_PREDICATE = `${oidcPrefix}tos_uri`;
export const POLICY_PREDICATE = `${oidcPrefix}policy_uri`;
export const CONTACTS_PREDICATE = `${oidcPrefix}contacts`;
export const LOGO_PREDICATE = `${oidcPrefix}logo_uri`;

export const APP_WEBID = "https://mockappurl.com";
export const APP_POLICY_URL = "https://mockappurl.com/privacy-policy";
export const APP_TOS_URL = "https://mockappurl.com/terms-of-service";

const contactThing = chain(
  createThing({ url: "https://mockappurl.com#contacts" }),
  (t) => addStringNoLocale(t, vcard.fn, "Example Contact"),
  (t) => addStringNoLocale(t, vcard.hasTelephone, "555-5555"),
  (t) => addStringNoLocale(t, vcard.hasEmail, "example@email.com"),
  (t) => addUrl(t, vcard.url, "https://examplewebsite.com")
);

const dataset = setThing(createSolidDataset(), contactThing);

export function mockApp() {
  return chain(
    createThing({ url: APP_WEBID }),
    (t) => addStringNoLocale(t, vcard.fn, "Mock App"),
    (t) => addUrl(t, vcardExtras("WebId"), APP_WEBID),
    (t) => addUrl(t, rdf.type, schema.SoftwareApplication), // just guessing what the type might be for apps
    (t) => addUrl(t, TOS_PREDICATE, APP_TOS_URL),
    (t) => addUrl(t, POLICY_PREDICATE, APP_POLICY_URL),
    (t) => addUrl(t, CONTACTS_PREDICATE, "https://mockappurl.com#contacts")
  );
}

export function mockAppDataset() {
  return setThing(dataset, mockApp());
}
