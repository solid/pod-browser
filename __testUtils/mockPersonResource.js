/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  asUrl,
  mockSolidDatasetFrom,
  mockThingFrom,
  setThing,
} from "@inrupt/solid-client";
import { vcard, foaf, rdf, space } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { packageProfile } from "../src/solidClientHelpers/profile";
import { vcardExtras } from "../src/addressBook";

const VCARD_WEBID_PREDICATE = vcardExtras("WebId");

export function mockWebIdNode(
  webId,
  nodeUrl = "https://example.org/contacts/Person/1234/index.ttl#4567"
) {
  const webIdNode = chain(
    mockThingFrom(nodeUrl),
    (t) => addUrl(t, rdf.type, VCARD_WEBID_PREDICATE),
    (t) => addUrl(t, vcard.value, webId)
  );
  const url = asUrl(webIdNode);
  return { webIdNode, webIdNodeUrl: url };
}

export const aliceProfileUrl = "http://alice.example.com/alice";
export const aliceWebIdUrl = "http://alice.example.com/alice#me";
export const aliceName = "Alice";
export const aliceNick = "A";
export const alicePhoto = "http://alice.example.com/alice.jpg";
export const alicePodRoot = "http://alice.example.com/";
export const aliceAlternativeProfileUrl = "https://alice2.example.org/card";
export const aliceAlternativeWebIdUrl = "https://alice2.example.org/card#me";

export function mockPersonThingAlice(...operations) {
  return chain(
    mockThingFrom(aliceWebIdUrl),
    (t) => addStringNoLocale(t, foaf.name, aliceName),
    (t) => addStringNoLocale(t, vcard.nickname, aliceNick),
    (t) => addUrl(t, vcard.hasPhoto, alicePhoto),
    (t) => addUrl(t, rdf.type, foaf.Person),
    (t) => addUrl(t, vcard.url, aliceAlternativeWebIdUrl),
    (t) => addUrl(t, space.storage, alicePodRoot),
    ...operations
  );
}

export function mockPersonDatasetAlice(...operations) {
  return setThing(
    mockSolidDatasetFrom(aliceWebIdUrl),
    mockPersonThingAlice(...operations)
  );
}

export function mockProfileAlice(...operations) {
  return packageProfile(aliceWebIdUrl, mockPersonDatasetAlice(...operations));
}

export const bobProfileUrl = "http://bob.example.com/bob";
export const bobWebIdUrl = "http://bob.example.com/bob#me";
export const bobName = "Bob";
export const bobNick = "B";
export const bobPodRoot = "http://bob.example.com/";
export const bobAlternateProfileUrl = "https://bob2.example.org/card";
export const bobAlternateWebIdUrl = "https://bob2.example.org/card#me";

export function mockPersonThingBob() {
  return chain(
    mockThingFrom(bobWebIdUrl),
    (t) => addStringNoLocale(t, foaf.name, bobName),
    (t) => addStringNoLocale(t, foaf.nick, bobNick),
    (t) => addUrl(t, rdf.type, foaf.Person),
    (t) => addUrl(t, foaf.openid, bobAlternateWebIdUrl),
    (t) => addUrl(t, space.storage, bobPodRoot)
  );
}

export function mockPersonDatasetBob() {
  return setThing(mockSolidDatasetFrom(bobWebIdUrl), mockPersonThingBob());
}

export function mockProfileBob() {
  return packageProfile(bobWebIdUrl, mockPersonDatasetBob());
}
