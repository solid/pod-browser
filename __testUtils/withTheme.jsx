/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
