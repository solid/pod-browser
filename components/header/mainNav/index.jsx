/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useContext, useState, useEffect } from "react";
import { createStyles, makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import {
  MainNav as PrismMainNav,
  MainNavLink,
} from "@inrupt/prism-react-components";
import Link from "next/link";
import { useRouter } from "next/router";
import styles from "./styles";
import FeatureContext from "../../../src/contexts/featureFlagsContext";
import getMainMenuItems from "../../../constants/mainMenu";

/* eslint react/jsx-props-no-spreading: off */

export const TESTCAFE_ID_MAIN_NAV = "main-nav";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));
const menuItems = getMainMenuItems();

function removeFeatureFlaggedItems(items, enabled) {
  return Promise.all(
    items.map(async ({ featureFlag, ...rest }) => {
      const active = !featureFlag || (await enabled(featureFlag));
      return {
        activePage: active,
        ...rest,
      };
    })
  );
}

export async function filterMenuLinks(items, enabled) {
  const links = await removeFeatureFlaggedItems(items, enabled);
  return links;
}

export default function MainNav() {
  const router = useRouter();
  const bem = useBem(useStyles());
  const { enabled } = useContext(FeatureContext);
  const [activeLinks, setActiveLinks] = useState([]);
  useEffect(() => {
    filterMenuLinks(menuItems, enabled).then((data) => setActiveLinks(data));
  }, [enabled]);
  const links = activeLinks
    .filter(({ activePage }) => activePage)
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    .map(({ featureFlag, activePage, path, label, pages, ...rest }) => {
      // extracting featureFlag and activePage props as we don't want them added to the html
      return {
        active: pages.includes(router.pathname),
        href: path,
        text: label,
        ...rest,
      };
    });

  return (
    <div className={bem("main-nav-container")}>
      <PrismMainNav
        links={links}
        data-testid={TESTCAFE_ID_MAIN_NAV}
        renderLink={(link) => (
          <Link href={link.href}>
            <MainNavLink {...link} />
          </Link>
        )}
      />
    </div>
  );
}
