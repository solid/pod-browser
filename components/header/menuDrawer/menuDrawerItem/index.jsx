/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint react/jsx-props-no-spreading:off */

import Link from "next/link";
import ListItem from "@material-ui/core/ListItem";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import { Icons } from "@inrupt/prism-react-components";
import ListItemText from "@material-ui/core/ListItemText";
import clsx from "clsx";
import React from "react";
import { createStyles, makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import T from "prop-types";
import styles from "./styles";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export default function MenuDrawerItem({
  href,
  icon,
  label,
  onClick,
  ...props
}) {
  const bem = useBem(useStyles());
  return href ? (
    <Link href={href} legacyBehavior>
      <ListItem
        key={label}
        onClick={onClick}
        className={bem("menu-drawer-item")}
        {...props}
      >
        <ListItemIcon className={bem("menu-drawer-item__icon-holder")}>
          <Icons name={icon} className={bem("menu-drawer-item__icon")} />
        </ListItemIcon>
        <ListItemText
          disableTypography
          primary={label}
          className={clsx(
            bem("user-menu__trigger"),
            bem("menu-drawer-item__text"),
            bem("menu-drawer-item__link")
          )}
        />
      </ListItem>
    </Link>
  ) : (
    <ListItem
      button
      key={label}
      onClick={onClick}
      className={bem("menu-drawer-item")}
      {...props}
    >
      <ListItemIcon className={bem("menu-drawer-item__icon-holder")}>
        <Icons name={icon} className={bem("menu-drawer-item__icon")} />
      </ListItemIcon>
      <ListItemText
        disableTypography
        primary={label}
        className={clsx(
          bem("user-menu__trigger"),
          bem("menu-drawer-item__text")
        )}
      />
    </ListItem>
  );
}

MenuDrawerItem.defaultProps = {
  href: null,
};

MenuDrawerItem.propTypes = {
  href: T.string,
  icon: T.string.isRequired,
  label: T.string.isRequired,
  onClick: T.func.isRequired,
};
