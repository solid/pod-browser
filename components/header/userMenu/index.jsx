/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { createStyles, makeStyles } from "@material-ui/styles";
import { UserMenu as PrismUserMenu } from "@inrupt/prism-react-components";
import { useBem } from "@solid/lit-prism-patterns";
import styles from "./styles";
import useAuthenticatedProfile from "../../../src/hooks/useAuthenticatedProfile";
import Spinner from "../../spinner";
import useUserMenu from "../../../src/hooks/useUserMenu";
import UserMenuAction from "./userMenuAction";

/* eslint react/jsx-props-no-spreading: off */

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

const TESTCAFE_ID_USER_MENU_BUTTON = "user-menu-button";
export const TESTCAFE_ID_USER_MENU = "user-menu";

export default function UserMenu() {
  const bem = useBem(useStyles());
  const authenticatedProfile = useAuthenticatedProfile();
  const menu = useUserMenu();

  if (!authenticatedProfile) return <Spinner />;

  return (
    <div className={bem("userMenu")} data-testid={TESTCAFE_ID_USER_MENU}>
      <PrismUserMenu
        className={bem("userMenu__trigger")}
        label={authenticatedProfile.name || "Account"}
        imgSrc={authenticatedProfile.avatar}
        menu={menu}
        menuId="UserMenu"
        data-testid={TESTCAFE_ID_USER_MENU_BUTTON}
        renderAction={(action) => <UserMenuAction {...action} />}
      />
    </div>
  );
}
