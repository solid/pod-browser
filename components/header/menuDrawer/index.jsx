/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint react/jsx-props-no-spreading:off */

import React, { useContext, useState, useEffect } from "react";
import { createStyles, makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import Drawer from "@material-ui/core/Drawer";
import Button from "@material-ui/core/Button";
import List from "@material-ui/core/List";
import Divider from "@material-ui/core/Divider";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import { Icons } from "@inrupt/prism-react-components";
import styles from "./styles";
import FeatureContext from "../../../src/contexts/featureFlagsContext";
import getMainMenuItems from "../../../constants/mainMenu";
import useUserMenu from "../../../src/hooks/useUserMenu";
import MenuDrawerItem from "./menuDrawerItem";
import { filterMenuLinks } from "../mainNav/index";

export const TESTCAFE_ID_MENU_DRAWER = "menu-drawer";
export const TESTCAFE_ID_MENU_DRAWER_BUTTON = "menu-drawer-button";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));
const menuItems = getMainMenuItems();

export default function MenuDrawer() {
  const bem = useBem(useStyles());
  const classes = useStyles();
  const { enabled } = useContext(FeatureContext);
  const [activeLinks, setActiveLinks] = useState([]);
  useEffect(() => {
    filterMenuLinks(menuItems, enabled).then((data) => setActiveLinks(data));
  }, [enabled]);
  const menuItemsToShow = activeLinks.filter(({ activePage }) => activePage);
  const [open, setOpen] = useState(false);
  const userMenu = useUserMenu();

  const handleDrawerOpen = () => setOpen(true);
  const handleDrawerClose = () => setOpen(false);

  return (
    <div
      className={classes.hamburgerMenu}
      data-testid={TESTCAFE_ID_MENU_DRAWER}
    >
      <Button
        onClick={handleDrawerOpen}
        data-testid={TESTCAFE_ID_MENU_DRAWER_BUTTON}
      >
        <Icons name="bars" aria-label="Menu" />
      </Button>
      <Drawer
        anchor="top"
        open={open}
        onClose={handleDrawerClose}
        classes={{ paper: classes.drawerContainer }}
      >
        <List>
          <ListItem button key="title">
            <ListItemText primary="Pod Browser" />
            <Button onClick={handleDrawerClose}>
              <Icons
                name="cancel"
                className={bem("icon-close")}
                aria-label="Pod Browser"
              />
            </Button>
          </ListItem>
          <Divider />
          {menuItemsToShow.map(({ path, ...action }) => (
            <MenuDrawerItem
              key={path}
              href={path}
              onClick={handleDrawerClose}
              {...action}
            />
          ))}
          <Divider />
          {userMenu.map(({ onClick, ...action }) => (
            <MenuDrawerItem
              key={action.label}
              onClick={() => {
                handleDrawerClose();
                onClick();
              }}
              {...action}
            />
          ))}
        </List>
      </Drawer>
    </div>
  );
}
