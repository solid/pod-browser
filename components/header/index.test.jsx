/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { useRouter } from "next/router";
import { waitFor } from "@testing-library/react";
import { renderWithTheme } from "../../__testUtils/withTheme";
import mockSessionContextProvider from "../../__testUtils/mockSessionContextProvider";
import Header, { TESTCAFE_ID_HEADER_LOGO } from "./index";
import {
  mockAuthenticatedSession,
  mockUnauthenticatedSession,
} from "../../__testUtils/mockSession";
import { TESTCAFE_ID_USER_MENU } from "./userMenu";
import { TESTCAFE_ID_MAIN_NAV } from "./mainNav";
import { TESTCAFE_ID_MENU_DRAWER } from "./menuDrawer";
import useAuthenticatedProfile from "../../src/hooks/useAuthenticatedProfile";
import { mockProfileAlice } from "../../__testUtils/mockPersonResource";

jest.mock("next/router");
const mockedRouterHook = useRouter;

jest.mock("../../src/hooks/useAuthenticatedProfile");
const mockedAuthenticatedHook = useAuthenticatedProfile;

describe("Header", () => {
  describe("with user logged in", () => {
    beforeEach(() => {
      mockedRouterHook.mockReturnValue({
        query: {
          iri: "https://mypod.myhost.com",
        },
      });
    });

    it("renders a header", async () => {
      const session = mockAuthenticatedSession();
      const SessionProvider = mockSessionContextProvider(session);
      mockedAuthenticatedHook.mockReturnValue(mockProfileAlice());

      const { asFragment, queryByTestId } = renderWithTheme(
        <SessionProvider>
          <Header />
        </SessionProvider>
      );
      await waitFor(() => {
        expect(queryByTestId(TESTCAFE_ID_HEADER_LOGO)).toBeDefined();
        expect(queryByTestId(TESTCAFE_ID_MENU_DRAWER)).toBeDefined();
        expect(queryByTestId(TESTCAFE_ID_MAIN_NAV)).toBeDefined();
        expect(queryByTestId(TESTCAFE_ID_USER_MENU)).toBeDefined();
      });
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("while user logged out", () => {
    it("renders only logo", () => {
      const session = mockUnauthenticatedSession();
      const SessionProvider = mockSessionContextProvider(session);

      const { asFragment, queryByTestId } = renderWithTheme(
        <SessionProvider>
          <Header />
        </SessionProvider>
      );

      expect(asFragment()).toMatchSnapshot();
      expect(queryByTestId(TESTCAFE_ID_HEADER_LOGO)).toBeDefined();
      expect(queryByTestId(TESTCAFE_ID_MENU_DRAWER)).toBeNull();
      expect(queryByTestId(TESTCAFE_ID_MAIN_NAV)).toBeNull();
      expect(queryByTestId(TESTCAFE_ID_USER_MENU)).toBeNull();
    });
  });
});
