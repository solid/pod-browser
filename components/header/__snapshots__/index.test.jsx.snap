// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header while user logged out renders only logo 1`] = `
<DocumentFragment>
  <header
    class="PodBrowser-header-banner"
  >
    <div
      class="PodBrowser-logoContainer"
    >
      <a
        data-testid="header-banner-logo"
        href="/"
      >
        <img
          alt="Inrupt PodBrowser"
          class="PodBrowser-header-banner__logo-image"
          height="40"
          src="/inrupt_logo-2020.svg"
        />
      </a>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Header with user logged in renders a header 1`] = `
<DocumentFragment>
  <header
    class="PodBrowser-header-banner"
  >
    <div
      class="PodBrowser-logoContainer"
    >
      <a
        data-testid="header-banner-logo"
        href="/"
      >
        <img
          alt="Inrupt PodBrowser"
          class="PodBrowser-header-banner__logo-image"
          height="40"
          src="/inrupt_logo-2020.svg"
        />
      </a>
    </div>
    <div
      class="PodBrowser-hamburgerMenu"
      data-testid="menu-drawer"
    >
      <button
        class="PodBrowser-root PodBrowser-root PodBrowser-text"
        data-testid="menu-drawer-button"
        tabindex="0"
        type="button"
      >
        <span
          class="PodBrowser-label"
        >
          <span
            aria-label="Menu"
            class="PodBrowser-icon-bars"
          />
        </span>
        <span
          class="PodBrowser-root"
        />
      </button>
    </div>
    <div
      class="PodBrowser-main-nav-container"
    >
      <nav
        class="PodBrowser-main-nav"
        data-testid="main-nav"
      >
        <ul
          class="PodBrowser-main-nav__list"
        >
          <li
            class="PodBrowser-main-nav__item"
          >
            <a
              class="PodBrowser-main-nav__link"
              data-testid="main-menu-files"
              href="/"
            >
              <span
                class="PodBrowser-icon-files PodBrowser-main-nav__icon"
              />
              <span
                class="PodBrowser-main-nav__label"
              >
                Files
              </span>
            </a>
          </li>
          <li
            class="PodBrowser-main-nav__item"
          >
            <a
              class="PodBrowser-main-nav__link"
              data-testid="main-menu-contacts"
              href="/contacts"
            >
              <span
                class="PodBrowser-icon-address-book PodBrowser-main-nav__icon"
              />
              <span
                class="PodBrowser-main-nav__label"
              >
                Contacts
              </span>
            </a>
          </li>
          <li
            class="PodBrowser-main-nav__item"
          >
            <a
              class="PodBrowser-main-nav__link"
              data-testid="main-menu-bookmarks"
              href="/bookmarks"
            >
              <span
                class="PodBrowser-icon-star PodBrowser-main-nav__icon"
              />
              <span
                class="PodBrowser-main-nav__label"
              >
                Bookmarks
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <div
      class="PodBrowser-userMenu"
      data-testid="user-menu"
    >
      <div
        class="PodBrowser-user-menu"
      >
        <button
          aria-controls="UserMenu"
          aria-expanded="false"
          aria-haspopup="true"
          class="PodBrowser-user-menu__trigger PodBrowser-userMenu__trigger"
          data-testid="user-menu-button"
          type="button"
        >
          <span
            class="PodBrowser-user-menu__trigger-profile-image"
            style="background-image: url(http://alice.example.com/alice.jpg);"
          >
            Â 
          </span>
          <span
            class="PodBrowser-user-menu__trigger-label"
          >
            Alice
          </span>
          <span
            class="PodBrowser-icon-caret-down PodBrowser-user-menu__trigger-chevron PodBrowser-user-menu__trigger-chevron--show"
          />
          <span
            class="PodBrowser-icon-caret-up PodBrowser-user-menu__trigger-chevron"
          />
        </button>
        <div
          aria-hidden="true"
          class="PodBrowser-user-menu__popup"
          id="UserMenu"
        >
          <ul
            class="PodBrowser-user-menu__list"
          >
            <li
              class="PodBrowser-user-menu__list-item"
            >
              <a
                class="PodBrowser-user-menu__list-item-trigger"
                data-testid="user-menu-profile"
              >
                <span
                  class="PodBrowser-icon-user PodBrowser-user-menu__list-item-icon"
                />
                <span
                  class="PodBrowser-user-menu__list-item-label"
                >
                  Profile
                </span>
              </a>
            </li>
            <li
              class="PodBrowser-user-menu__list-item"
            >
              <button
                class="PodBrowser-user-menu__list-item-trigger"
                data-testid="user-menu-logout"
                type="button"
              >
                <span
                  class="PodBrowser-icon-log-out PodBrowser-user-menu__list-item-icon"
                />
                <span
                  class="PodBrowser-user-menu__list-item-label"
                >
                  Log out
                </span>
              </button>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
